### ISSUE TITLE
AWS Layer handler serious performance issue when application import certain libraries

### ISSUE THREAD

**Issue by zhaowb** [getsentry/sentry-python#1736]
### How do you use Sentry?

Sentry Saas (sentry.io)

### Version

1.10.1

### Steps to Reproduce

1. Create empty aws lambda function

File: lambda_function.py
```python
# import hubspot  # https://github.com/HubSpot/hubspot-api-python
def lambda_handler(event, context):
    pass
```
File: mysentry.py
```python
# mysentry.py
# https://docs.sentry.io/platforms/python/guides/aws-lambda/
import os
import sentry_sdk
from sentry_sdk.integrations.aws_lambda import AwsLambdaIntegration

sentry_sdk.init(
    # dsn="https://examplePublicKey@o0.ingest.sentry.io/0",
    dsn=os.environ['SENTRY_DSN'],
    # https://github.com/getsentry/sentry-python/blob/a5ee1bd8c5b456704b9629fc430fb5203602f3c7/sentry_sdk/consts.py#L56
    environment=os.environ['SENTRY_ENVIRONMENT'],
    integrations=[
        AwsLambdaIntegration(),
    ],
    # Set traces_sample_rate to 1.0 to capture 100%
    # of transactions for performance monitoring.
    # We recommend adjusting this value in production,
    traces_sample_rate=float(os.environ['SENTRY_TRACES_SAMPLE_RATE']),
)
module_name, handler_fn_name = os.environ['SENTRY_INITIAL_HANDLER'].rsplit('.', 1)
import importlib
handler_module = importlib.import_module(module_name)
handler_fn = getattr(handler_module, handler_fn_name)

def handler(event, context):
    return handler_fn(event, context)
```

2. Add layer (2 versions tested, both have same issue)
arn:aws:lambda:ap-southeast-2:943013980633:layer:SentryPythonServerlessSDK:24
arn:aws:lambda:ap-southeast-2:943013980633:layer:SentryPythonServerlessSDK:42

3. Set handler (3 handlers to compare performance)
lambda_function.lambda_handler  # default empty handler function
sentry_sdk.integrations.init_serverless_sdk.sentry_lambda_handler  # sentry handler to enable sentry
mysentry.handler  # this shows init sentry is ok, the problem is in layer implementation

4. Set env
SENTRY_DSN="https://examplePublicKey@o0.ingest.sentry.io/0"
SENTRY_ENVIRONMENT=production
SENTRY_INITIAL_HANDLER=lambda_function.lambda_handler
SENTRY_TRACES_SAMPLE_RATE=0.1

5. Test following situations, total 6 combinations
- with/without import hubspot
- test each handler (with any layer, same result)

### Expected Result

Sentry init should not take too long

### Actual Result

The combination "import hubspot" + "handler=sentry_lambda_handler" takes ~30 seconds to finish with 128MB mem. 1GB runs faster (~3 seconds).

The combination "import hubspot" + "handler=lambda_handler" takes ~2.3 seconds to finish with 128MB mem.
This shows the problem should not happend even with lowest mem.

The combination "import hubspot" + "handler=mysentry.handler" takes ~3.2 seconds to finish with 128MB mem.
This shows enable sentry sdk doesn't have the problem. The problem is in layer implementation.

The combination "no import" + "handler=lambda_handler" takes ~0.1 seconds with 128MB mem.
This shows a baseline of performance.

**Comment by zhaowb** [comment]
`hubspot-api-python` uses a lot of auto generated code and does many dynamic build at import time. Maybe that's causing the issue or make it worse enough to be noticeable. 
My point is the sentry layer handler should do same thing as sentry sdk init as suggested in official document. The layer is just for less code change.

**Comment by antonpirker** [comment]
Hey @zhaowb !
Thanks for reporting this. 
The Sentry layer does not only call `sentry_sdk.init()` but also starts a rust binary (called Sentry `relay`). So maybe this is somehow interfering with hubspot.

Can you time how long the combination "no import" + "handler=sentry_lambda_handler" would take?

**Comment by zhaowb** [comment]
@antonpirker 

When no import, with 128MB

"handler=lambda_handler" takes ~0.1 seconds

With sentry layer version 24,

"handler=sentry_lambda_handler" takes ~1.3 seconds
"handler=mysentry.handler" takes ~1.3 seconds

With sentry layer version 42,

"handler=sentry_lambda_handler" takes ~1.8 seconds
"handler=mysentry.handler" takes ~1.8 seconds.

When use layer 42, both handler sentry_lambda_handler and mysentry can see relay server started in log. This means the relay server might not be the reason.

eg following is layer 42, handler=mysentry.handler, import hubspot

```
Function Logs
+ exec /opt/relay/relay run --mode=proxy --shutdown-timeout=2 --aws-runtime-api=127.0.0.1:9001
2022-11-14T18:43:11Z [relay::setup] INFO: launching relay without config folder
2022-11-14T18:43:11Z [relay::setup] INFO:   relay mode: proxy
2022-11-14T18:43:11Z [relay::setup] INFO:   relay id: -
2022-11-14T18:43:11Z [relay::setup] INFO:   public key: -
2022-11-14T18:43:11Z [relay::setup] INFO:   log level: INFO
2022-11-14T18:43:11Z [relay_server::actors::upstream] INFO: upstream relay started
2022-11-14T18:43:11Z [relay_server::actors::outcome] INFO: Configured to emit outcomes as client reports
2022-11-14T18:43:11Z [relay_server::actors::outcome] INFO: OutcomeProducer started.
2022-11-14T18:43:11Z [relay_server::actors::processor] INFO: starting 2 envelope processing workers
2022-11-14T18:43:11Z [relay_server::actors::envelopes] INFO: envelope manager started
2022-11-14T18:43:11Z [relay_server::actors::relays] INFO: key cache started
2022-11-14T18:43:11Z [relay_server::actors::project_cache] INFO: project cache started
2022-11-14T18:43:11Z [relay_aws_extension::aws_extension] INFO: AWS extension started
2022-11-14T18:43:11Z [relay_aws_extension::aws_extension] INFO: Registering AWS extension on http://127.0.0.1:9001/2020-01-01/extension
2022-11-14T18:43:11Z [relay_aws_extension::aws_extension] INFO: AWS extension successfully registered
2022-11-14T18:43:11Z [relay_server::service] INFO: spawning http server
2022-11-14T18:43:11Z [relay_server::service] INFO:   listening on: http://127.0.0.1:3000/
2022-11-14T18:43:11Z [actix_net::server::server] INFO: Starting 2 workers
2022-11-14T18:43:11Z [actix_net::server::server] INFO: Starting server on 127.0.0.1:3000
2022-11-14T18:43:11Z [relay_system::controller] INFO: relay server starting
2022-11-14T18:43:11Z [relay_server::actors::project_local] INFO: project local cache started
2022-11-14T18:43:11Z [relay_server::actors::project_upstream] INFO: project upstream cache started
2022-11-14T18:43:11Z [relay_server::actors::outcome_aggregator] INFO: outcome aggregator started
2022-11-14T18:43:11Z [relay_metrics::aggregation] INFO: aggregator started
sentry_sdk.VERSION 1.10.1
EXTENSION	Name: sentry-lambda-extension	State: Ready	Events: [INVOKE,SHUTDOWN]
START RequestId: 4813c56e-75ed-4221-845f-1aec95edd19b Version: $LATEST
END RequestId: 4813c56e-75ed-4221-845f-1aec95edd19b
REPORT RequestId: 4813c56e-75ed-4221-845f-1aec95edd19b	Duration: 263.21 ms	Billed Duration: 264 ms	Memory Size: 128 MB	Max Memory Used: 128 MB	Init Duration: 3118.34 ms
```

**Comment by zhaowb** [comment]
@antonpirker no worries, I'm not blocked by this issue.
I'm planning to create a layer my own with just pre-install sentry package and the mysentry.py as above, so it would be a drop-in replacement of the official layer.
Since using sentry_sdk.init() is suggested in [official doc](https://docs.sentry.io/platforms/python/guides/aws-lambda/), I suppose I won't lose much function comparing to the official layer, except I need to maintain the package version by my own.

**Comment by sentrivana** [comment]
Closing since relay is not part of the AWS Lambda layer anymore.
